File: src/dotnet/Ansight.Tools.Preferences/RemovePreferenceKeyTool.cs
namespace Ansight.Tools.Preferences;

public sealed class RemovePreferenceKeyTool : ITool
{
    private readonly PreferencesToolsOptions options;

    public RemovePreferenceKeyTool(PreferencesToolsOptions? options = null)
    {
        this.options = options ?? PreferencesToolsOptions.Default;
    }

    public string Category => "prefs";

    public ToolPolicy Policy => ToolPolicy.Critical;

    public string Id => PreferencesToolIds.RemoveKey;

    public string Name => "Remove Preference Key";

    public string Description => "Deletes a key from shared preferences or user defaults.";

    public string Keywords => "preferences sharedpreferences nsuserdefaults storage delete remove";

    public ToolSchema ArgumentsSchema => PreferencesToolSchemas.RemoveKeyArguments;

    public ToolSchema ResultSchema => PreferencesToolSchemas.RemoveKeyResult;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(PreferencesSupport.RemoveKey(options, arguments)));
        }
        catch (PlatformNotSupportedException exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "prefs_platform_unsupported"));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "prefs_remove_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsPreferences/RemovePreferenceKeyTool.swift
import AnsightKit
import Foundation

public final class RemovePreferenceKeyTool: AnsightTool {
    private let options: AnsightPreferencesToolOptions

    public init(options: AnsightPreferencesToolOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightPreferencesToolIds.removeKey,
            name: "Remove Preference Key",
            description: "Deletes a key from an iOS UserDefaults store.",
            category: "prefs",
            scope: AnsightToolScope.delete.rawValue,
            keywords: "preferences userdefaults delete remove settings",
            security: AnsightPreferencesToolSecurityProfiles.removeKey,
            argumentsSchema: AnsightPreferencesToolSchemas.removeKeyArguments,
            resultSchema: AnsightPreferencesToolSchemas.removeKeyResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightPreferencesSupport.removeKey(options: options, arguments: arguments))
        } catch {
            return .failure(error.localizedDescription, errorCode: "prefs_remove_failed")
        }
    }
}
